### PR TITLE
Configuring the VR Editor for standalone devices

### PR DESCRIPTION
**Pull request by m4gr3d**
https://user-images.githubusercontent.com/914968/210125651-3b15f4a8-ba1d-4218-8723-51d523676de6.mp4

### Generate VR Editor apks:
* Meta:
  * Dev build
  ```
  godot/platform/android/java/gradlew assembleOpenxrMetaDev
  ```
   * Release build
  ```
  godot/platform/android/java/gradlew assembleOpenxrMetaDebug
  ```
* Pico:
  * Dev build
  ```
  godot/platform/android/java/gradlew assembleOpenxrPicoDev
  ```
  * Release build
  ```
  godot/platform/android/java/gradlew assembleOpenxrPicoDebug
  ```

### Known issues
* Controllers are not being tracked

### CODE REVIEW

File: platform/android/display_server_android.cpp [hunk 1]
```diff
@@ -658,16 +658,19 @@ void DisplayServerAndroid::cursor_set_custom_image(const Ref<Resource> &p_cursor
...
 DisplayServer::VSyncMode DisplayServerAndroid::window_get_vsync_mode(WindowID p_window) const {
 #if defined(VULKAN_ENABLED)
-	return context_vulkan->get_vsync_mode(p_window);
-#else
-	return DisplayServer::VSYNC_ENABLED;
+	if (rendering_driver == "vulkan") {
+		return context_vulkan->get_vsync_mode(p_window);
+	}
 #endif
+	return DisplayServer::VSYNC_ENABLED;
```
Comment (BastiaanOlij) [inline platform/android/display_server_android.cpp:677]: Hmm, is GLES ignoring VSync all together ?

File: platform/android/java/editor/build.gradle [hunk 2]
```diff
@@ -76,25 +76,86 @@ android {
...
+            }
+            applicationIdSuffix ".openxr"
+            versionNameSuffix "-openxr"
+        }
+
+        google {
+            dimension "xrRuntime"
+        }
+        meta {
+            dimension "xrRuntime"
```
Comment (BastiaanOlij) [inline platform/android/java/editor/build.gradle:89]: Hmm, we need to add Khronos in here as well, still need to add the Khronos loader to the godot xr loader repo but I have no device to test this on.

**Comment by BastiaanOlij** [comment]
I don't think we can include the loader .so into the repo. Godot has a strict no binaries rule, everything must be source but shouldn't the Godot XR loaders AAR be used here as well?

**Comment by m4gr3d** [comment]
> I don't think we can include the loader .so into the repo. Godot has a strict no binaries rule, everything must be source but shouldn't the Godot XR loaders AAR be used here as well?

The Android editor is a full APK so it already includes closed binaries (the Gradle dependencies). 
The Godot XR loaders AARs are just wrappers around the shared libraries, so they're unnecessary in a regular Android project.

**Comment by m4gr3d** [comment]
At the APK build time, all the dependencies including the shared libraries are required, so the options are either:
- include the shared libraries directly
- turn the openxr loaders into maven dependencies and add them to build.gradle. The shared libraries are technically not part of the source but they're downloaded and included at build time
- use a git submodule to include the shared libraries

File: platform/android/java/editor/build.gradle (hunk 2 above)
Comment (m4gr3d) [reply platform/android/java/editor/build.gradle:89]: What device does the khronos loader run on?

The changes target the editor so if we don't plan to support/polish the experience on a particular set of devices, we omit the inclusion of their dependencies.

**Comment by BastiaanOlij** [comment]
> * turn the openxr loaders into maven dependencies and add them to build.gradle. The shared libraries are technically not part of the source but they're downloaded and included at build time

This I think would make the most sense.

**Comment by BastiaanOlij** [comment]
Hopefully some day both Pico and Meta support the Khronos loader and we can just compile the source thats already present in `thirdparty` and not deal with external dependencies, but that is future music.

File: platform/android/java/editor/build.gradle (hunk 2 above)
Comment (BastiaanOlij) [reply platform/android/java/editor/build.gradle:89]: HTC Focus I think, and on normal Android devices if you use Monado as an XR Runtime.

**Comment by m4gr3d** [comment]
> > * turn the openxr loaders into maven dependencies and add them to build.gradle. The shared libraries are technically not part of the source but they're downloaded and included at build time
> 
> This I think would make the most sense.

Done!

File: platform/android/java/editor/build.gradle [hunk 3]
```diff
@@ -76,26 +68,106 @@ android {
...
+    implementation libraries.androidxFragment
+    implementation project(":lib")
+
+    implementation "androidx.window:window:1.0.0"
+
+    // Meta OpenXR dependencies
+    openxrMetaImplementation "io.github.m4gr3d:godot-openxr-loaders-meta:1.0.0"
+
+    // Pico OpenXR dependencies
+    openxrPicoImplementation "io.github.m4gr3d:godot-openxr-loaders-pico:1.0.0"
```
Comment (m4gr3d) [inline platform/android/java/editor/build.gradle]: @BastiaanOlij See https://github.com/GodotVR/godot_openxr_loaders/pull/18 for the logic to enable pushing the libraries to mavencentral

**Comment by m4gr3d** [comment]
cc @rsjtdrjgfuzkfg, can you validate if the build works on Pico devices, I don't have access to one for testing.

**Comment by rsjtdrjgfuzkfg** [comment]
> cc @rsjtdrjgfuzkfg, can you validate if the build works on Pico devices, I don't have access to one for testing.

The patch is not sufficient to start the editor on my Pico 4, assuming I built the right thing.

I built b2a7a5c6cbc8fb210bea0cda1ffc2109720e7fdd with

    scons platform=android target=template_release arch=arm64
    scons platform=android target=template_debug arch=arm64
    scons platform=android target=template_debug arch=arm64 dev_build=True
    scons platform=android target=editor arch=arm64
    scons platform=android target=editor arch=arm64 dev_build=True
    ( cd platform/android/java ; ./gradlew --no-daemon assembleOpenxrPicoDebug )
    ( cd platform/android/java ; ./gradlew --no-daemon assembleOpenxrPicoDev )

The resulting apks do switch to VR mode, ask for storage permission and initialize the Pico loader, but don't exit the loading animation (the animation comes form the OS, it is always displayed until an application starts rendering; this indicates that Godot never starts rendering). From logcat:

    I GodotPluginRegistry: Initializing Godot plugin GodotOpenXRPico
    I GodotPluginRegistry: Completed initialization for Godot plugin GodotOpenXRPico
    W System.err: java.io.FileNotFoundException: _cl_
    [... stacktrace for not finding a command line ...]
    E godot   : USER ERROR: Couldn't load file '/project.binary', error code 12.
    E godot   :    at: _load_settings_text_or_binary (core/config/project_settings.cpp:746)
    I godot   : Godot Engine v4.0.beta.custom_build.b2a7a5c6c - https://godotengine.org
    I godot   : OpenXR: XR_KHR_loader_init_android is not reported as available - trying to initialize anyway...
    I PxrLoader: Info [GENERAL | xrCreateInstance | PxrLoader] : LoaderInstance::CreateInstance succeeded with 0 layers enabled and runtime interface - created instance = [... pointer ...]
    I godot   : OpenXR: Running on OpenXR runtime:  Pico(XRT) by Pico et al '02bf2b8'   3.0.1
    I godot   : OpenGL Renderer: Adreno (TM) 650
    I GodotInputHandler: === New Input Device: pvr-virtual-input-0
    [... some more input handlers ...]
    I godot   : OpenXR: Requested OpenGL version exceeds the maximum version this runtime has been tested on and is known to support.
    I godot   : - desired_version  3.3.0
    I godot   : - minApiVersionSupported  2.0.0
    I godot   : - maxApiVersionSupported  3.2.1023
    W GodotIO : Shared storage access is limited on Android 10 and higher.
    [... the warning regarding shared storage is repeated throughout ...]
    [... messages from the Pico runtime that the activity is resumed ...]
    I godot   : OpenXR: No viewport was marked with use_xr, there is no rendered output!


You can likely ignore two of the warnings:
* `OpenXR: XR_KHR_loader_init_android is not reported as available - trying to initialize anyway...` is a regression from https://github.com/godotengine/godot/pull/69392, the check happens before extensions are listed so the warning is always printed, although Pico does advertise the extension.
* The warning regarding the OpenGL version also occurs when starting games that use the OpenGL renderer (and games do start to render, exiting the loading animation).

**Comment by m4gr3d** [comment]
> but don't exit the loading animation

@rsjtdrjgfuzkfg How long did you wait for the loading animation? Performance is still slow and so it takes a while for the first frame to start rendering.

On the Quest Pro, it took ~30s for the `dev` build and ~20s for the `debug` build before it started rendering.
It's an area to start looking into as soon as the input issue is resolved.

**Comment by BastiaanOlij** [comment]
It's still required to start with the `--xrmode on` switch isn't it? From the logs its showing it isn't turning `use_xr` mode on on the viewport though it wouldn't be giving that error if it wasn't starting OpenXR..

**Comment by m4gr3d** [comment]
> It's still required to start with the `--xrmode on` switch isn't it? From the logs its showing it isn't turning `use_xr` mode on on the viewport though it wouldn't be giving that error if it wasn't starting OpenXR..

That switch is added by the [`GodotXREditor` class](https://github.com/BastiaanOlij/godot/pull/18/files#diff-e5dd87e230d850c4a9085305c6d7c45b36cb5cddfa55524a6174d023f6e7d446) on start-up.

@rsjtdrjgfuzkfg The apk has two `ACTION_MAIN` activities, so another area to check is whether the right activity is launched on your device (i.e: should be running `GodotXRProjectManager`).

**Comment by rsjtdrjgfuzkfg** [comment]
With the debug build, I waited about a minute and left the device running in standby for a bit, checking in after a few tens of minutes. So I think I waited long enough.

Regarding the activity: there is only a launcher icon for `org.godotengine.editor.GodotProjectManager`, so Pico's launcher might not properly handle apps with multiple main actions. Manually starting the correct activity (`adb shell am start org.godotengine.editor.v4.openxr.pico/org.godotengine.editor.GodotXRProjectManager`) has the same issue, though (same output applies, though the messages printed by the system now refer to `GodotXRProjectManager` as activity, so the launch seems to have worked).

**Pull request merged**